ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java

```diff
@@ -0,0 +1,278 @@
+  }
+
+  @Override
+  public void addDataNodeConfigChangeListener(DataNodeConfigChangeListener listener) throws Exception {
+    Subscription subscription = new Subscription(listener);
+    propertyStore.subscribe(CONFIG_PATH, subscription);
+    subscription.start();
+  }
+
+  @Override
+  public boolean set(DataNodeConfig config) {
+    ZNRecord record = converter.convert(config);
+    String path = CONFIG_PATH + "/" + record.getId();
+    return propertyStore.set(path, record, AccessOption.PERSISTENT);
+  }
+
+  @Override
+  public DataNodeConfig get(String instanceName) {
+    String path = CONFIG_PATH + "/" + instanceName;
+    ZNRecord record = propertyStore.get(path, new Stat(), AccessOption.PERSISTENT);
+    return record != null ? converter.convert(record) : null;
+  }
+
+  private class Subscription implements HelixPropertyListener {
+    private final DataNodeConfigChangeListener listener;
+    private final CompletableFuture<Void> initFuture = new CompletableFuture<>();
+
+    Subscription(DataNodeConfigChangeListener listener) {
+      this.listener = listener;
+    }
+
+    /**
+     * Tell the background executor to perform any initialization needed. This should be called after registering this
+     * listener using {@link HelixPropertyStore#subscribe}. This will ensure that the first event sent to
+     * {@link #listener} will contain the entire set of current configs.
+     * @throws Exception if there was an exception while making the initialization call.
+     */
+    void start() throws Exception {
+      eventExecutor.execute(this::initializeIfNeeded);
+      try {
+        initFuture.get();
+      } catch (ExecutionException e) {
+        throw Utils.extractExecutionExceptionCause(e);
+      }
+    }
+
+    @Override
+    public void onDataChange(String path) {
+      LOGGER.debug("DataNodeConfig path {} changed", path);
+      onPathChange(path);
+    }
+
+    @Override
+    public void onDataCreate(String path) {
+      LOGGER.debug("DataNodeConfig path {} created", path);
+      onPathChange(path);
+    }
+
+    @Override
+    public void onDataDelete(String path) {
+      // TODO handle node deletions dynamically. Doing so requires further work in the ClusterChangeHandler impl
+      LOGGER.info("DataNodeConfig path {} deleted. This requires a restart to handle", path);
+    }
+
+    private void onPathChange(String path) {
+      eventExecutor.execute(() -> updateOrInitialize(path));
+    }
+
+    private Iterable<DataNodeConfig> lazyIterable(Collection<ZNRecord> records) {
+      return () -> records.stream().map(converter::convert).filter(Objects::nonNull).iterator();
+    }
+
+    /**
+     * Either perform a full initialization or notify the listener about an incremental update to a path. This should be
+     * called from the event executor thread.
+     * @param changedPath the path that changed.
+     */
+    private synchronized void updateOrInitialize(String changedPath) {
```

Comment:
This comment might be out of scope but I wonder if node (frontend/server) lost connection to ZK and got reconnected, is there a mechanism to re-read all datanode configs in PropertyStore (and re-populate in-mem clustermap)?
As far as I know, InstanceConfig current has an async way to read all configs again (not ideal) but in next release of Helix, it's changed to a sync way, which guarantees all configs will be read again after reconnection.

Reply:
That is a good point, since you could potentially miss updates while zk was disconnected. I will follow up with Helix regarding this. If it is not possible for this to happen with the PropertyStore impl, we could consider using a background polling thread that periodically fetches all configs.